[SPARK-4012] stop SparkContext when the exception is thrown from an infinite loop

CodingCat opened the pull request:
https://issues.apache.org/jira/browse/SPARK-4012

This patch is a resubmission for https://github.com/apache/spark/pull/2864

What I am proposing in this patch is that **_when the exception is thrown from an infinite loop, we should stop the SparkContext, instead of let JVM throws exception forever**_

So, in the infinite loops where we originally wrapped with a `logUncaughtExceptions`, I changed to `tryOrStopSparkContext`, so that the Spark component is stopped

Early stopped JVM process is helpful for HA scheme design, for example, 

The user has a script checking the existence of the pid of the Spark Streaming driver for monitoring the availability; with the code before this patch, the JVM process is still available but not functional when the exceptions are thrown

@andrewor14, @srowen , mind taking further consideration about the change?

CodingCat committed:
exit JVM process when the exception is thrown from an infinite loop

SparkQA commented:
[Test build #28525 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28525/consoleFull) for   PR 5004 at commit [`6322959`](https://github.com/apache/spark/commit/63229590d4645e0da8cb7e4217931a0443f7900d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28525 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28525/consoleFull) for   PR 5004 at commit [`6322959`](https://github.com/apache/spark/commit/63229590d4645e0da8cb7e4217931a0443f7900d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28525/
Test PASSed.

zsxwing commented:
Do you think calling `Thread.setDefaultUncaughtExceptionHandler(SparkUncaughtExceptionHandler)` in the driver side is acceptable? Or at least, I think we should make sure every thread created by Spark should set `SparkUncaughtExceptionHandler`.

CodingCat commented:
Hi, @zsxwing , `Thread.setDefaultUncaughtExceptionHandler(SparkUncaughtExceptionHandler)` is fine, though it's equivalent to the current patch in terms of functionality 

but I'm not sure if we need to set exceptionHandler of all threads as SparkUncaughtExceptionHandler, because that means, once there is an uncaught exception, we stop the program...

zsxwing commented:
> once there is an uncaught exception, we stop the program...

Yes. So I suggests only threads created by Spark should set SparkUncaughtExceptionHandler. I think if a Spark internal thread throws an uncaught exception, it often means some Spark internal module has crashed. Just my 2 cents about improving the robustness.

Of course, look good to me about your changes.

CodingCat commented:
@zsxwing , sounds reasonable

let's wait for more eyes on this

srowen commented:
I tend to favor this change for `ContextCleaner` and `AsynchronousListenerBus`, because an exception there kills a thread that is clearly intended to run forever. I'm not as sure about `FsHistoryProvider`, since there it is not as clear (yet) that the `Runnable` should cause the worker to die if it hits an exception. why this one?

I'd like it if, say, @aarondav might comment, since he added one of the lines being changed.

CodingCat commented:
@srowen thanks for the comments, 

the reason to change FsHistoryProvider is that the runner generated by this function  is essentially executed by the threads with fixed rate

https://github.com/CodingCat/spark/blob/SPARK-4012-1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L138

aarondav commented:
Is it really OK to System.exit() the driver JVM? This may be user code that has an embedded SparkContext. The SparkUncaughtExceptionHandler is suitable for Executors, where we have full control over the JVM, and AppClient for the same reason, but I'm not sure TaskSchedulerImpl should be using it, or ContextCleaner, for instance.

CodingCat committed:
stop SparkContext instead of quit the JVM process

CodingCat committed:
revise comments

SparkQA commented:
[Test build #28613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28613/consoleFull) for   PR 5004 at commit [`6087864`](https://github.com/apache/spark/commit/6087864c7b1e0558627234aabc3d85c7e830cf59).
 * This patch merges cleanly.

CodingCat commented:
@aarondav , thanks for the insightful suggestion

I just updated the patch

the change becomes a bit bigger, as I need to create a new method in Utils as `tryOrStopSparkContext`, which is a curried method in the current patch

also, I need to pass a SparkContext reference to `AsynchronousListenerBus` to be able to pass sparkContext as the parameter, so I changed the `start()` to `start(sparkContext: SparkContext)` in which I set the value of the newly added sparkContext variable member of  `AsynchronousListenerBus`

SparkQA commented:
[Test build #28613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28613/consoleFull) for   PR 5004 at commit [`6087864`](https://github.com/apache/spark/commit/6087864c7b1e0558627234aabc3d85c7e830cf59).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28613/
Test PASSed.

aarondav commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1156,6 +1156,18 @@ private[spark] object Utils extends Logging {
   }
 
   /**
+   * Execute a block of code that evaluates to Unit, stop SparkContext is any uncaught exception
```
Add a comment contrasting this to tryOrExit, saying that this method is suitable for the driver while tryOrExit should be used for other JVMs started by Spark, over which we have full control. Also, second part should say something like "stopping the SparkContext if there is any uncaught exception."

aarondav commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1156,6 +1156,18 @@ private[spark] object Utils extends Logging {
+      block
+    } catch {
+      case e: ControlThrowable => throw e
+      case t: Throwable => sc.stop()
```
Add a null check around here, just to protect tests and whatnot

aarondav replied:
Also add an error log printing the exception and the reason we're stopping sc.

aarondav commented on core/src/main/scala/org/apache/spark/util/AsynchronousListenerBus.scala:
```diff
@@ -90,8 +93,9 @@ private[spark] abstract class AsynchronousListenerBus[L <: AnyRef, E](name: Stri
    * This should only be called once.
    */
-  def start() {
+  def start(sc: SparkContext) {
```
Add an `@param sc Used to stop the SparkContext in case the listener thread dies.`, or something to that effect

aarondav commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -145,7 +145,7 @@ private[spark] class TaskSchedulerImpl(
       sc.env.actorSystem.scheduler.schedule(SPECULATION_INTERVAL milliseconds,
             SPECULATION_INTERVAL milliseconds) {
-        Utils.tryOrExit { checkSpeculatableTasks() }
+        Utils.tryOrStopSparkContext(sc) { checkSpeculatableTasks() }
```
Thanks for updating this one as well!

aarondav commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -93,7 +93,7 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
   private def getRunner(operateFun: () => Unit): Runnable = {
     new Runnable() {
-      override def run() = Utils.logUncaughtExceptions {
+      override def run() = Utils.tryOrExit {
```
I am not sure what the FsHistoryProvider is, what should the failure semantics be around it?

CodingCat replied:
FsHistoryProvider is the module used by history server process to render the application logs....

this getRunner is called for periodically check updated logs and clean logs...

the point here is if anything uncaught is thrown, like OOM, HistoryServer process should be restarted, since it has been unfunctional...

CodingCat committed:
address the comments

CodingCat commented:
@aarondav thanks for the comments, I just updated the patch

SparkQA commented:
[Test build #28619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28619/consoleFull) for   PR 5004 at commit [`3c72cd8`](https://github.com/apache/spark/commit/3c72cd8bb3237454e0866deb274d80db1d07f897).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28619/consoleFull) for   PR 5004 at commit [`3c72cd8`](https://github.com/apache/spark/commit/3c72cd8bb3237454e0866deb274d80db1d07f897).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28619/
Test PASSed.

aarondav commented:
Alright, this looks good to me. @srowen and @zsxwing, any comments?

zsxwing commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1156,6 +1158,28 @@ private[spark] object Utils extends Logging {
+          logError(s"uncaught error in thread ${Thread.currentThread().getName}, stopping " +
+            "SparkContext", t)
+          sc.stop()
+        }
```
How about throwing `t` again here? So that the user can use `UncaughtExceptionHandler` to monitor the uncaught exception. If not, the user won't be aware that `sc` is shutdown. until calling `runJob` next time.

CodingCat replied:
Hi, @zsxwing thanks for the comments

I personally prefer a more conservative way here (the current approach)

Because the throwable thrown from here can be varying in terms of types, and I'm concerning that the Throwable from here, like `OOM`, would be mixed with the instances of the same type generated by the other components in user's program; on the other hand, our goal is just to let the user know `SparkContext` is stopped

So I prefer to letting the user call SparkContext.runJob to get a 'new IllegalStateException("SparkContext has been shutdown")' which (hopefully) will be handled `exactly`

@srowen @aarondav , your thoughts?

aarondav replied:
What if we catch NonFatal(e) and re-throw other Throwables? Basically saying that fatal errors should be re-thrown, but lesser ones can just stop here, they should only application-level exceptions which are our code's concern.

srowen commented:
I don't have a strong opinion on this one. I suppose there's a question of what may happen if you stop the `SparkContext` but then immediately kill the calling thread with an uncaught `Throwable`. Does that undermine the action of `stop()` in some cases?

CodingCat commented:
@srowen I checked the code, one suspicious part might be the asynchronous shutdown of ActorSystem, but that should be OK in Spark case...

@aarondav ?

zsxwing commented:
Since `tryOrStopSparkContext` is in Utils, I think it's better to check if it's reasonable in the general case. I'm concerning that swallowing fatal errors (OOM) or `InterruptException` may cause some issue.

CodingCat committed:
throw fatal error again

SparkQA commented:
[Test build #28825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28825/consoleFull) for   PR 5004 at commit [`589276a`](https://github.com/apache/spark/commit/589276a496e7371de1e51b37b7afdf4af7c27286).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28825/consoleFull) for   PR 5004 at commit [`589276a`](https://github.com/apache/spark/commit/589276a496e7371de1e51b37b7afdf4af7c27286).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28825/
Test PASSed.

CodingCat commented:
@zsxwing @srowen @aarondav how about the current version?

zsxwing commented:
LGTM

aarondav commented:
LGTM

CodingCat commented:
thanks, guys~

aarondav commented:
Cool, merging this into master. Thanks!
